src/grafana_agent.py

```diff
@@ -119,6 +113,39 @@ def __init__(self, *args):
             self._on_loki_push_api_endpoint_departed,
         )
 
+    # Abstract Methods
+
+    def agent_version_output(self) -> str:
```

Comment:
Hate to say this now, but now that I see how the code looks like with this pattern, I kind of think I'd like it more if we used a different pattern. 
Instead of abstract charm + inheritance, I'd go for concrete charm + composition + two separate grafana controller classes.

Why:
- charm code is complicated enough without throwing inheritance into the mix. Suddenly we have three charms to think about.
- I liked the pattern in loki https://github.com/canonical/loki-k8s-operator/blob/main/src/loki_server.py where we have a separate LokiServer class. 

So I'd rather have a single charm class (no subclasses) handling the operation logic, and sending out calls to a GrafanaServer (or whatever) class, that is either a GrafanaServerK8s or a GrafanaServerMachine.
So the charm code becomes simpler:
```
if self.grafana_server.is_ready(): ...
```

We can also say: too late, that ship has sailed, and I'll shut up and review this PR instead.

Reply:
Both solutions seem to be more or less the same to me and I'd rather not rewrite it. We can discuss further with the whole team.